### PR TITLE
Rust implementations to replace data tables (#161)

### PR DESCRIPTION
**Pull request by peterjoel**
A few points to discuss, as it's my first PR to this project:

I introduced a dependency on Primal. I couldn't see a better alternative and could never get the same speed without it.

My tests look at the highest and lowest values found in the original hard-coded data tables, with extra cases on the step boundaries. The functions will support much higher input values, which I didn't search for corresponding test values for, since they would only be needed 20 years in the future.

I wasn't sure about which constants to use versus introducing new ones. For example the python code I borrowed from Ethereum had HASH_BYTES = 64, while this project had a constant NODE_BYTES with the same value. I'm not confident enough of what these values actually mean to know if it would be correct to use them to mean the same thing.

The codebase seems to have a lot of files with lines ending in whitespace. My editor is correcting those (I think this is what is happening at least), so my commits are riddled with unnecessary whitespace changes. I'm not 100% sure how to prevent this, short of manually modifying all my commits afterwards. I left them in for now, so you can see. For comparison, the Servo project does not permit trailing whitespace at all, which prevents this issue.

### CODE REVIEW

File: ethash/src/compute.rs [hunk 1]
```diff
@@ -34,6 +34,10 @@ pub const ETHASH_MIX_BYTES: usize = 128;
 pub const ETHASH_ACCESSES:usize =  64;
```
Comment (gavofyork) [inline ethash/src/compute.rs]: double space.

File: ethash/src/compute.rs [hunk 2]
```diff
@@ -34,6 +34,10 @@ pub const ETHASH_MIX_BYTES: usize = 128;
...
 pub const ETHASH_DATASET_PARENTS:u32 = 256;
 
+const DATASET_BYTES_INIT :u64 = 1<<30;
```
Comment (gavofyork) [inline ethash/src/compute.rs]: missing space around `<<`. `:` should have space following, not preceeding.

File: ethash/src/compute.rs [hunk 3]
```diff
@@ -34,6 +34,10 @@ pub const ETHASH_MIX_BYTES: usize = 128;
 pub const ETHASH_ACCESSES:usize =  64;
 pub const ETHASH_DATASET_PARENTS:u32 = 256;
```
Comment (gavofyork) [inline ethash/src/compute.rs]: space should follow `:`

File: ethash/src/compute.rs [hunk 4]
```diff
@@ -34,6 +34,10 @@ pub const ETHASH_MIX_BYTES: usize = 128;
...
 
+const DATASET_BYTES_INIT :u64 = 1<<30;
+const DATASET_BYTES_GROWTH : u64 = 1<<23;
+const CACHE_BYTES_INIT: u64 = 1<<24;
```
Comment (gavofyork) [inline ethash/src/compute.rs]: missing space around `<<`.

File: ethash/src/compute.rs [hunk 5]
```diff
@@ -34,6 +34,10 @@ pub const ETHASH_MIX_BYTES: usize = 128;
...
+const CACHE_BYTES_GROWTH: u64 = 1<<17;
```
Comment (gavofyork) [inline ethash/src/compute.rs]: missing space around `<<`.

File: ethash/src/compute.rs [hunk 6]
```diff
@@ -149,14 +153,22 @@ fn sha3_512(input: &[u8], output: &mut [u8]) {
 
 #[inline]
 fn get_cache_size(block_number: u64) -> usize {
-	assert!(block_number / ETHASH_EPOCH_LENGTH < 2048);
-	return CACHE_SIZES[(block_number / ETHASH_EPOCH_LENGTH) as usize] as usize;
+    let mut sz:u64 = CACHE_BYTES_INIT + CACHE_BYTES_GROWTH * (block_number / ETHASH_EPOCH_LENGTH);
+    sz = sz - NODE_BYTES as u64;
+    while ! is_prime( sz / NODE_BYTES as u64 ) {
```
Comment (gavofyork) [inline ethash/src/compute.rs]: additional space following `!`, after `(`, before `)`

File: ethash/src/compute.rs [hunk 7]
```diff
@@ -149,14 +153,22 @@ fn sha3_512(input: &[u8], output: &mut [u8]) {
...
+        sz  = sz - 2 * NODE_BYTES as u64;
+    }
+    sz as usize
 }
 
 #[inline]
 fn get_data_size(block_number: u64) -> usize {
-	assert!(block_number / ETHASH_EPOCH_LENGTH < 2048);
-	return DAG_SIZES[(block_number / ETHASH_EPOCH_LENGTH) as usize] as usize;
+    let mut sz:u64 = DATASET_BYTES_INIT + DATASET_BYTES_GROWTH * (block_number / ETHASH_EPOCH_LENGTH);
```
Comment (gavofyork) [inline ethash/src/compute.rs]: missing space after `:`

File: ethash/src/compute.rs [hunk 8]
```diff
@@ -149,14 +153,22 @@ fn sha3_512(input: &[u8], output: &mut [u8]) {
...
+    sz = sz - ETHASH_MIX_BYTES as u64;
+    while ! is_prime(sz / ETHASH_MIX_BYTES as u64){
```
Comment (gavofyork) [inline ethash/src/compute.rs]: additional space following `!`
missing space between `)` and `{`.

File: ethash/src/compute.rs [hunk 9]
```diff
@@ -322,9 +334,34 @@ fn to_hex(bytes: &[u8]) -> String {
 }
 
 #[test]
+fn test_get_cache_size() {
+	// https://github.com/ethereum/wiki/wiki/Ethash/ef6b93f9596746a088ea95d01ca2778be43ae68f#data-sizes
+	assert_eq!( 16776896usize, get_cache_size(0));
```
Comment (gavofyork) [inline ethash/src/compute.rs]: additional space after `(`; also on following lines.

File: ethash/src/compute.rs [hunk 10]
```diff
@@ -322,9 +334,34 @@ fn to_hex(bytes: &[u8]) -> String {
...
+	assert_eq!( 16776896usize, get_cache_size(1));
+	assert_eq!( 16776896usize, get_cache_size(ETHASH_EPOCH_LENGTH-1));
```
Comment (gavofyork) [inline ethash/src/compute.rs]: missing spaces around `-`.

File: ethash/src/compute.rs [hunk 11]
```diff
@@ -322,9 +334,34 @@ fn to_hex(bytes: &[u8]) -> String {
...
+	assert_eq!( 16907456usize, get_cache_size(ETHASH_EPOCH_LENGTH));
+	assert_eq!( 16907456usize, get_cache_size(ETHASH_EPOCH_LENGTH+1));
```
Comment (gavofyork) [inline ethash/src/compute.rs]: missing spaces around `+`.

File: ethash/src/compute.rs [hunk 12]
```diff
@@ -322,9 +334,34 @@ fn to_hex(bytes: &[u8]) -> String {
...
+	assert_eq!( 285081536usize, get_cache_size(2047 * ETHASH_EPOCH_LENGTH));
+	assert_eq!( 285081536usize, get_cache_size(2048 * ETHASH_EPOCH_LENGTH - 1));
+}
+
+#[test]
+fn test_get_data_size() {
+	// https://github.com/ethereum/wiki/wiki/Ethash/ef6b93f9596746a088ea95d01ca2778be43ae68f#data-sizes
+	assert_eq!( 1073739904usize, get_data_size(0));
+	assert_eq!( 1073739904usize, get_data_size(1));
+	assert_eq!( 1073739904usize, get_data_size(ETHASH_EPOCH_LENGTH-1));
```
Comment (gavofyork) [inline ethash/src/compute.rs]: missing spaces around `-`.

File: ethash/src/compute.rs [hunk 13]
```diff
@@ -322,9 +334,34 @@ fn to_hex(bytes: &[u8]) -> String {
...
+	assert_eq!( 1082130304usize, get_data_size(ETHASH_EPOCH_LENGTH));
+	assert_eq!( 1082130304usize, get_data_size(ETHASH_EPOCH_LENGTH+1));
```
Comment (gavofyork) [inline ethash/src/compute.rs]: missing spaces around `+`.

**Comment by gavofyork** [comment]
logic and tests look reasonable. more attention needed to consistency of whitespace.

**Comment by tomusdrw** [comment]
@peterjoel Try running rustfmt with those settings: https://github.com/ethcore/parity/blob/a8045ed37ea09d6650e871fdfd35836ec84ab67a/rustfmt.toml
should fix all whitespace issues.

**Comment by arkpar** [comment]
looks good to me

**Comment by gavofyork** [comment]
@ethcore-cla-bot help

**Comment by gavofyork** [comment]
[clabot:check]

**Comment by peterjoel** [comment]
Where can I find the CLA? The bot itself doesn't provide any links and there is no mention of it in the project's README.

**Comment by gavofyork** [comment]
sorry about that - clabot appears to have misconfigured itself.

link is https://cla.ethcore.io/

**Comment by peterjoel** [comment]
[clabot:check]

**Pull request merged**